lib/rules/consistent-this.js

```diff
@@ -9,36 +10,79 @@
 //------------------------------------------------------------------------------
 
 module.exports = function(context) {
+    var alias = context.options[0];
 
-    return {
+    /**
+     * Checks that an assignment to an identifier only assigns 'this' to the
+     * appropriate alias, and the alias is only assigned to 'this'.
+     * @param {ASTNode} node The assigning node.
+     * @param {String} name The name of the identifier.
+     * @param {ASTNode} value The expression to be assigned.
+     * @returns {void}
+     */
+    function checkAssignment(node, name, value) {
+        var isThis = value !== null && value.type === "ThisExpression";
 
-        "VariableDeclaration": function(node) {
-            var alias = context.options[0];
+        if (name === alias) {
+            if (!isThis) {
+                context.report(node, "Designated alias '{{alias}}' is not assigned to 'this'.", { alias: name });
+            }
+        } else if (isThis) {
+            context.report(node, "Unexpected alias '{{alias}}' for 'this'.", { alias: name });
+        }
+    }
 
-            node.declarations.forEach(function(declaration) {
-                if (declaration.id.name === alias &&
-                        !(declaration.init && declaration.init.type === "ThisExpression")) {
 
-                    context.report(
-                        node,
-                        "Designated 'this' alias '{{alias}}' is not assigned " +
-                            "to the current execution context.",
-                        { alias: declaration.id.name }
-                    );
-                }
+    /**
+     * Checks that a set of variable declarations correctly defines the 'this'
+     * alias, and does not assign 'this' to a different declaration. If the
+     * declaration declares a 'this' alias, but does not assign to it, it is
+     * required that the alias is assigned to 'this' later in the same scope.
+     * @param {ASTNode} node The variable declaration node.
+     * @returns {void}
+     */
+    function checkVariableDeclaration(node) {
+        node.declarations.forEach(function (decl) {
+            var body, child, i;
+
+            if (decl.id.name === alias && decl.init === null) {
+                body = node.parent.body;
+
+                for (i = body.indexOf(node) + 1; i < body.length; i++) {
```

Comment:
This is doing a traversal, which is inefficient because the rule already does a traversal. It would be better to watch for entering/exiting functions and identifier references to see if `this` is assigned.